homeassistant/components/mqtt/client.py

```diff
@@ -911,7 +921,9 @@ async def _discovery_cooldown(self) -> None:
             await asyncio.sleep(wait_until - now)
             now = time.time()
             last_discovery = self._mqtt_data.last_discovery
-            last_subscribe = self._last_subscribe
+            last_subscribe = (
+                now if self._pending_subscriptions else self._last_subscribe
```

Comment:
I think we need to check for `self._pending_subscriptions` also in the while condition, maybe:
```py
while now < wait_until or self._pending_subscriptions:
```